pulp_ansible/app/sigstore/tasks/sigstore_signature.py

```diff
@@ -0,0 +1,295 @@
+        """Initialize Sigstore signing first stage."""
+        super().__init__()
+        self.content = content
+        self.sigstore_signing_service = sigstore_signing_service
+        self.repos_current_sigstore_signatures = current_sigstore_signatures
+        self.semaphore = asyncio.Semaphore(settings.ANSIBLE_SIGNING_TASK_LIMITER)
+
+    async def sigstore_sign_collection_versions(self, collection_versions):
+        """Signs the collection versions with Sigstore."""
+
+        # Prepare ephemeral key pair and certificate and sign the collections asynchronously
+        private_key = ec.generate_private_key(ec.SECP384R1())
+        client_secret = self.sigstore_signing_service.oidc_client_secret
+        identity_token = self.sigstore_signing_service.issuer.identity_token(
+            "sigstore", client_secret, self.sigstore_signing_service.enable_interactive
+        )
+        oidc_identity = Identity(identity_token)
+
+        # Build an X.509 Certificiate Signing Request
+        builder = (
+            x509.CertificateSigningRequestBuilder()
+            .subject_name(
+                x509.Name(
+                    [
+                        x509.NameAttribute(NameOID.EMAIL_ADDRESS, oidc_identity.proof),
+                    ]
+                )
+            )
+            .add_extension(
+                x509.BasicConstraints(ca=False, path_length=None),
+                critical=True,
+            )
+        )
+        certificate_request = builder.sign(private_key, hashes.SHA256())
+        certificate_response = self.sigstore_signing_service.fulcio.signing_cert.post(
+            certificate_request, identity_token
+        )
+
+        # Verify the SCT
+        sct = certificate_response.sct
+        cert = certificate_response.cert
+        chain = certificate_response.chain
+
+        verify_sct(sct, cert, chain, self.sigstore_signing_service.rekor._ct_keyring)
+
+        # Limits the number of subprocesses spawned/running at one time
+        async with self.semaphore:
+            async for collection_version in collection_versions:
+                # We create the checksums manifest to sign
+                async with aiofiles.tempfile.NamedTemporaryFile(
+                    dir=".", mode="w", delete=False
+                ) as manifest_file:
+                    manifest_data = await sync_to_async(_generate_checksum_manifest)(
+                        collection_version
+                    )
+                    await manifest_file.write(manifest_data)
+                async with aiofiles.open(manifest_file.name, mode="rb", buffering=0) as iofile:
+                    async with aiofiles.tempfile.NamedTemporaryFile(
+                        dir=".", mode="w", delete=False
+                    ) as manifest_content:
+                        content = await iofile.read()
+                        manifest_content.write(content)
+                    with open(manifest_content.name, "rb") as manifest_bytes:
```

Comment:
Same two things as above with the verify code:
1. Can we just extract the `MANIFEST.IN` file instead of regenerating it?
2. Once we have the file in memory let's use io.BytesIO instead of rewriting it to a temp file.

Reply:
Just use [TextIo](https://docs.python.org/3/library/io.html#text-i-o) instead.